saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/AbstractSaml2AuthenticationRequest.java

```diff
@@ -89,6 +98,24 @@ public String getAuthenticationRequestUri() {
 		return this.authenticationRequestUri;
 	}
 
+	/**
+	 * Returns the id of this AuthNRequest.
+	 * @return the id of this request
+	 * @since 5.7
+	 */
+	public String getId() {
```

Comment:
Thinking about this more, if the only place the id is needed is in OpenSAML-specific classes, then it's reasonable to leave this property off and have `OpenSaml4AuthenticationProvider` parse the `AuthnRequest` using OpenSAML and read the ID. This serves the function of keeping the public API as small as possible.

Reply:
Hmm... can't we just store the `AuthnRequest` instead of the serialized message string in `AbstractSaml2AuthenticationRequest`? Can't really find anywhere where this string is used for anything particular and in the tests where it is used, it's deserialized into an `AuthnRequest` so it would make more sense to store the actual request... what do you think?

Reply:
I mean of course the request string is used but what do you say about storing the `AuthnRequest` and then add a method `getSerializedSamlRequest()` to `AbstractSaml2AuthenticationRequest` which returns the serialized request for use in `SamlWebSsoAuthenticationRequestFilter` (so we move the serialization process to reside inside `AbstractSaml2AuthenticationRequest`). My concern is otherwise that the deserialization of the request in `OpenSaml4AuthenticationProvider` will be a bit hacky and cumbersome and to me, it makes more sense to store the domain class of the request instead of the serialized string :)

Reply:
`AbstractSaml2AuthenticationRequest` cannot contain an `AuthnRequest` because `AbstractSaml2AuthenticationRequest` is not OpenSAML-specific.

Reply:
> My concern is otherwise that the deserialization of the request in OpenSaml4AuthenticationProvider will be a bit hacky

It's not clear to me why this would be hacky. It does need to be serialized so it can go into the session. Whether it's Java or it's Spring, something is thus going to have to deserialize it.